### PR TITLE
Add sample for resequencer 

### PR DESCRIPTION
**Pull request by LakshanWeerasinghe**
## Purpose
$subject

### CODE REVIEW

File: resequencer/resquencer.bal [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+};
+
+map<Approval[]> approvalsMap = {};
+
+final http:Client hrClient = check new ("http://api.wso2hr.com.balmock.io");
+
+service /api/v1 on new http:Listener(8080) {
+
+    resource function post employees/[string emoloyeeId]/approval(@http:Payload Approval approval) returns error? {
+        if approvalsMap.hasKey(emoloyeeId) {
```
Comment (manuranga) [inline resequencer/resquencer.bal]: refactor so the if dosn't cover everything.

File: resequencer/resquencer.bal [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            Approval[] approvals = approvalsMap.get(emoloyeeId);
+            approvals.push(approval);
+            if approvals.length() == 3 {
+                from Approval orderedApproval in reOrderLeadsApproval(approvals)
+                    do {
+                        json _ = check hrClient->/promotions/employees/[emoloyeeId]/approvals.post(orderedApproval);
+                    };
+                _ = approvalsMap.remove(emoloyeeId);
+                return;
```
Comment (manuranga) [inline resequencer/resquencer.bal]: get rid `return` by using `else`.

File: resequencer/resquencer.bal [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+final http:Client hrClient = check new ("http://api.wso2hr.com.balmock.io");
+
+service /api/v1 on new http:Listener(8080) {
+
+    resource function post employees/[string emoloyeeId]/approval(@http:Payload Approval approval) returns error? {
+        if approvalsMap.hasKey(emoloyeeId) {
+            Approval[] approvals = approvalsMap.get(emoloyeeId);
+            approvals.push(approval);
+            if approvals.length() == 3 {
+                from Approval orderedApproval in reOrderLeadsApproval(approvals)
```
Comment (manuranga) [inline resequencer/resquencer.bal]: Lets try to have a single query.

File: resequencer/resquencer.bal (hunk 1 above)
Comment (LakshanWeerasinghe) [reply resequencer/resquencer.bal]: Updated.

File: resequencer/resquencer.bal [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            approvalsMap[emoloyeeId] = [approval];
+            return;
+        }
+        Approval[] approvals = approvalsMap.get(emoloyeeId);
+        approvals.push(approval);
+        if approvals.length() == 3 {
+            from Approval orderedApproval in approvals
+                order by orderedApproval.leadLevel
+                do {
+                    json _ = check hrClient->/promotions/employees/[emoloyeeId]/approval.post(orderedApproval);
```
Comment (ShammiL) [inline resequencer/resquencer.bal]:
```suggestion
                    _ = check hrClient->/promotions/employees/[emoloyeeId]/approval.post(orderedApproval, targetType = json);
```

File: resequencer/resequencer.yml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+tagline: Reorder a sequence of messages recived
+desc: Resequencer re order the sequence of messages recived in the incorrect order.
```
Comment (ShammiL) [inline resequencer/resequencer.yml]:
```suggestion
desc: Resequencer reorders the sequence of messages received in the incorrect order.
```

File: resequencer/resquencer.bal [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+const TEAM_LEAD = 3;
+
+map<Approval[]> approvalsMap = {};
+
+final http:Client hrClient = check new ("http://api.wso2hr.com.balmock.io");
+
+service /api/v1 on new http:Listener(8080) {
+
+    resource function post employees/[string emoloyeeId]/approval(Approval approval) returns error? {
+        if !approvalsMap.hasKey(emoloyeeId) {
```
Comment (manuranga) [inline resequencer/resquencer.bal]: having both `hasKey` and `get` are duplicate work.

File: resequencer/resquencer.bal [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+            approvalsMap[emoloyeeId] = [approval];
+            return;
+        }
+        Approval[] approvals = approvalsMap.get(emoloyeeId);
+        approvals.push(approval);
+        if approvals.length() == 3 {
+            from Approval orderedApproval in approvals
+                order by orderedApproval.leadLevel
```
Comment (manuranga) [inline resequencer/resquencer.bal]: align.

File: resequencer/resquencer.bal [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+map<Approval[]> approvalsMap = {};
+
+final http:Client hrClient = check new ("http://api.wso2hr.com.balmock.io");
+
+service /api/v1 on new http:Listener(8080) {
+
+    resource function post employees/[string emoloyeeId]/approval(Approval approval) returns error? {
+        Approval[]? approvals = approvalsMap[emoloyeeId];
+        if approvals == () {
+            approvalsMap[emoloyeeId] = [approval];
```
Comment (manuranga) [inline resequencer/resquencer.bal]: let's use the early return style for both `if`s

File: resequencer/resquencer.bal [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        } else {
+            approvals.push(approval);
+            if approvals.length() == 3 {
+                from Approval orderedApproval in approvals
```
Comment (manuranga) [inline resequencer/resquencer.bal]: `approval`

File: resequencer/resquencer.bal [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+type Approval record {
+    SENIOR_MANAGER|MANAGER|TEAM_LEAD leadLevel;
+    int score;
+    int nextPromotionLevel;
+};
+
+const SENIOR_MANAGER = 1;
+const MANAGER = 2;
+const TEAM_LEAD = 3;
```
Comment (manuranga) [inline resequencer/resquencer.bal]: let's change the order around. and move this above `Approval`

File: resequencer/resquencer.bal [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+    resource function post employees/[string emoloyeeId]/approval(Approval approval) returns error? {
+        Approval[]? approvals = approvalsMap[emoloyeeId];
+        if approvals == () {
+            approvalsMap[emoloyeeId] = [approval];
+            return;
+        }
+        approvals.push(approval);
+        if approvals.length() < 3 {
+            approvalsMap[emoloyeeId] = approvals;
```
Comment (manuranga) [inline resequencer/resquencer.bal]: can we get rid of this line?

File: resequencer/resquencer.bal [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    TEAM_LEAD|MANAGER|SENIOR_MANAGER leadLevel;
+    int score;
+    int nextPromotionLevel;
+};
+
+const TEAM_LEAD = 1;
+const MANAGER = 2;
+const SENIOR_MANAGER = 3;
+
+map<Approval[]> approvalsMap = {};
```
Comment (manuranga) [inline resequencer/resquencer.bal]: let's call this `partialApprovals` or `incompleteApprovals`

File: resequencer/resquencer.bal [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+        Approval[]? approvals = approvalsMap[emoloyeeId];
+        if approvals == () {
+            approvalsMap[emoloyeeId] = [approval];
+            return;
+        }
+        approvals.push(approval);
+        if approvals.length() < 3 {
+            return;
+        }
+        from Approval appro in approvals
```
Comment (manuranga) [inline resequencer/resquencer.bal]: `appro` is not a good abbreviation. ether `a` or `approval`

**Review by manuranga — APPROVED** [review]

**Pull request merged**